src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/AbstractAddItemCommand.cs

```diff
@@ -67,19 +68,53 @@ protected override Task<CommandStatusResult> GetCommandStatusAsync(IProjectTree
         {
             var nodeToAddTo = GetNodeToAddTo(node);
 
-            // Call OnAddingNodesAsync.
-            // Then publish changes that could have taken place in OnAddingNodesAsync.
-            await OnAddingNodesAsync(nodeToAddTo).ConfigureAwait(true);
-            await _projectTree.TreeService.PublishLatestTreeAsync(waitForFileSystemUpdates: true).ConfigureAwait(true);
-
-            var updatedNode = _projectTree.CurrentTree.Find(node.Identity);
-            var updatedNodeToAddTo = _projectTree.CurrentTree.Find(nodeToAddTo.Identity);
-            var addedNodes = OrderingHelper.GetAddedNodes(nodeToAddTo, updatedNodeToAddTo);
+            var addedElements = await OrderingHelper.AddItems(_projectVsServices.ActiveConfiguredProject, _projectVsServices.ProjectLockService, () => OnAddingNodesAsync(nodeToAddTo)).ConfigureAwait(true);
 
-            if (addedNodes.Any())
+            if (addedElements.Any())
             {
-                // Make sure to change ConfigureAwait to "true" if we need switch back.
-                await OnAddedNodesAsync(_projectVsServices.ActiveConfiguredProject, node, addedNodes, updatedNode).ConfigureAwait(false);
+                await new ProjectAccessor(_projectVsServices.ProjectLockService).OpenProjectForWriteAsync(_projectVsServices.ActiveConfiguredProject, project => OnAddedElements(project, addedElements, node, nodeToAddTo)).ConfigureAwait(true);
+
+                // Re-select the node that was the target.
+                await _projectTree.TreeService.PublishLatestTreeAsync(waitForFileSystemUpdates: true).ConfigureAwait(true);
+                HACK_NodeHelper.Select(_projectVsServices.ActiveConfiguredProject, _serviceProvider, node);
+
+                // If the node we wanted to add to is a folder, make sure it is expanded.
+                if (nodeToAddTo.IsFolder)
+                {
+                    HACK_NodeHelper.ExpandFolder(_projectVsServices.ActiveConfiguredProject, _serviceProvider, nodeToAddTo);
+                }
+
+                await Task.Delay(1000).ConfigureAwait(true);
+
+                if (addedElements.Length == 1)
+                {
+                    var unconfiguredProject = _projectVsServices.ActiveConfiguredProject.UnconfiguredProject;
+                    var filePath = unconfiguredProject.MakeRooted(addedElements[0].Include);
+                    var addedNode = _projectTree.CurrentTree.FindImmediateChildByPath(filePath);
+
+                    Assumes.NotNull(addedNode);
+
+                    var uiShellOpenDocument = _serviceProvider.GetService(typeof(SVsUIShellOpenDocument)) as IVsUIShellOpenDocument;
```

Comment:
Also, you may want to pull this functionality into `IVsShellUtilitiesHelper` (an existing MEF export in your solution). Then inside that you can use `VsShellUtilities.OpenDocument` or `OpenDocumentWithSpecificEditor`. This will greatly reduce the number of arguments you need to pass to it. This is how CPS does it: https://mseng.visualstudio.com/VSIDEProj/_git/VSIDEProj.CPS?path=%2Fsrc%2FMicrosoft.VisualStudio.ProjectSystem.VS.Implementation%2FShell%2FVsShellUtilitiesWrapper.cs&version=GBmaster&_a=contents

Reply:
This is most likely going to be removed. The reason I wanted to open a document was due to a bug we found; but there may be a better way to handle this.